Add DSF and DSDIFF file types management

damien78 opened the pull request:
This adds the management of tags for the DSD audio files types: DSF and DSDIFF
- DSF: backport from the work done by sbooth on the taglib2 branch, ID3v2 tags as per the DSF standard
- DSDIFF: manages 'DIIN' chunk (Edited Master Information chunk, for title and artist) as per the DSDIFF 1.5 standard, and adds management of an 'ID3 ' (ID3v2) chunk as found in popular softwares understanding DSDIFF that store it at either the root level, or for some as a child of the 'PROP' chunk

sbooth commented:
This looks good to me.  I think it would be beneficial to add at least some minimal tests for these formats, though.  I can see about whipping some up if @damien78 doesn't have time.

FestusHagen commented on taglib/dsf/dsffile.cpp:
```diff
@@ -0,0 +1,231 @@
+
+  // File is malformed or corrupted
+  if(d->fileSize != length()) {
+    printf("DSF::File::read() -- File is corrupted wrong length");
```
`printf()` needs to be replaced with `debug()`

FestusHagen commented:
All the `typedef` types (`uint`, `ulong`, etc...) need to be changed to standard types.

FestusHagen commented:
Many extra spaces, formatting inconsistencies, need to run astyle on it ... or do it manually!

FestusHagen commented on taglib/CMakeLists.txt:
```diff
@@ -364,8 +382,17 @@ set_target_properties(tag PROPERTIES
+      VERSION "A"
+      SOVERSION "A"
+      )
 endif()
```
Personally, I believe the BUILD_FRAMEWORK changes should be in a separate pull request.

damien78 commented:
Updated (last commit) the pull request to change the remaining typedef types to standard ones, and remove the BUILD_FRAMEWORK changes.

@sbooth : which tests do you have in mind in addition to the ones I've made (test_dsdiff.cpp). For test_dsf.cpp, I've just copied yours.

FestusHagen commented:
@damien78, There are many many formatting issues and inconsistencies within itself let alone with TagLib style.

For example:
- Spaces between function names and opening parentheses. (some with, some without)
- Stray spaces.
- Blank lines with whitespace.
- Inconsistent blank lines within code.
- Inconsistent blank lines between functions.
- Inconsistently placed curly braces.
- Inconsistent curly brace use. (Single line blocks with and without curly braces)
- Inconsistently placed else statements.
- Inconsistent indentation.
- Operators lacking spaces around them. (foo+1 vs. foo + 1)
- Inconsistent inline if statements. (if(!isValid()) return;) (Not TagLib style)
- Inconsistent copyright headers.

I know this sounds picky, it is, contributors (mostly @TsudaKageyu) have spent a lot of time cleaning up this project, not saying it's 100%, close though ... so lets work consistent, forward!

Please refer to Astyle or TagLib code for TagLib style. (Or ask)

damien78 commented:
Ok, so is there a tool to do this formatting as automatically as possible?

TsudaKageyu commented:
@damien78 Great work! I like it. However, I have a few things to point out.

* Use `String()` instead of `String::null`. `String::null` is now obsolete.
* Add tests for your file types to `test_fileref.cpp` since you added them to `FileRef`.

sbooth commented:
@damien78 I overlooked the DSDIFF tests, sorry. I think the core test functionality is there.

An option for code formatting is [astyle](http://astyle.sourceforge.net), and TagLib has https://github.com/taglib/taglib/blob/master/.astylerc

FestusHagen commented:
@TsudaKageyu does that go for `ByteVector::null` as well?

FestusHagen commented:
@damien78, I had surgery last Friday, so as I lay here twiddling my thumbs studying your contribution I found myself formatting/styling it, So I decided to push it to my repo, you are welcome to take
advantage of my boredom if you so choose!

There are no operational changes, in fact I'd bet you could diff the processed Assembly and find no difference.

I do not guarantee it is perfectly formatted/styled, close though. I am human.

I don't wrap, Thus I don't know wrap, Therefore it's wrapped the way it was wrapped!

TsudaKageyu commented:
Right, I had to mention `ByteVector::null`, @FestusHagen. I have amended my comment.

damien78 commented:
Just updated (last commit) to replace the deprecated ::null, though this is not strictly equivalent, as String() returns false with .isNull(), so needs to be tested also with .isEmpty()
And I've added the DSF and DSDIFF tests to test_fileref.cpp
And thanks @FestusHagen ! I've directly copied your changes. I hope your are recovering fast...

FestusHagen commented on taglib/dsdiff/dsdiffdiintag.cpp:
```diff
@@ -0,0 +1,162 @@
+
+void DSDIFF::DIIN::Tag::setTitle(const String &title)
+{
+  if (title.isNull() || title.isEmpty())
```
Whoops ... I Missed a space between function name and parenthesis ... 

And I thought I was perfect!

FestusHagen commented on taglib/dsdiff/dsdiffdiintag.cpp:
```diff
@@ -0,0 +1,162 @@
+
+void DSDIFF::DIIN::Tag::setArtist(const String &artist)
+{
+  if (artist.isNull() || artist.isEmpty())
```
And another missed space between function name and parenthesis ...

FestusHagen commented on taglib/dsdiff/dsdifffile.cpp:
```diff
@@ -0,0 +1,761 @@
+  }
+
+  // Couldn't find an existing chunk, so let's create a new one.
+  unsigned int i =  d->chunks.size() - 1;
```
Here I missed a stray space between `=  d`.

FestusHagen commented on taglib/dsdiff/dsdifffile.cpp:
```diff
@@ -0,0 +1,761 @@
+  }
+  if(d->childChunkIndex[DIINChunk] >= static_cast<int>(startingChunk)) {
+    std::vector<Chunk64> &childChunksToUpdate = d->childChunks[DIINChunk];
+    if(childChunksToUpdate.size()>0) {
```
Here I forgot spaces around an operator ... Should be `if(childChunksToUpdate.size() > 0) {`

FestusHagen commented on taglib/dsdiff/dsdifffile.cpp:
```diff
@@ -0,0 +1,761 @@
+  if(!isValid())
+    return;
+
+  if(d->childChunkIndex[PROPChunk]<0) {
```
Again, spaces around an operator ... Should be `if(d->childChunkIndex[PROPChunk] < 0) {`

FestusHagen commented on taglib/dsdiff/dsdifffile.cpp:
```diff
@@ -0,0 +1,761 @@
+          d->tag.access<DSDIFF::DIIN::Tag>(DIINIndex, false)->setTitle(titleStr);
+        }
+      }
+      else if (d->childChunks[DIINChunk][i].name == "DIAR") {
```
Again, I missed a space between function name and parenthesis.

FestusHagen commented:
Thank you @damien78, fortunately I do heal fast, unfortunately it's 3 weeks flat no matter!
1 down 2 to go.

I wouldn't do a commit just for my 6 errors indicated by my line notes above, that's just me publicly flogging myself ... :(

I did notice you have two different copyright dates on your work, is that what you want ...

Looks good to me ... Thank you for hanging in there!

-Enjoy

FestusHagen commented on taglib/toolkit/tfile.cpp:
```diff
@@ -149,6 +151,10 @@ PropertyMap File::properties() const
     return dynamic_cast<const ASF::File* >(this)->properties();
+  if (dynamic_cast<const DSF::File* >(this))
+    return dynamic_cast<const DSF::File* >(this)->properties();
+  if (dynamic_cast<const DSDIFF::File* >(this))
```
I didn't even look at existing TagLib files modified with this PR.
Shame on me!

Thus these  two places have a space between function name and parenthesis ...

FestusHagen commented on taglib/toolkit/tfile.cpp:
```diff
@@ -178,6 +184,10 @@ void File::removeUnsupportedProperties(const StringList &properties)
     dynamic_cast<ASF::File* >(this)->removeUnsupportedProperties(properties);
+  else if (dynamic_cast<DSF::File* >(this))
+    dynamic_cast<DSF::File* >(this)->removeUnsupportedProperties(properties);
+  else if (dynamic_cast<DSDIFF::File* >(this))
```
And these  two places have a space between function name and parenthesis ...

FestusHagen commented:
I'm sorry @damien78, I didn't even look at existing TagLib files modified by this PR thus missed the above line noted locations in `toolkit/tfile.cpp` ...

Those I would like to see fixed and committed, So I guess they all should be ...

Maybe it's me that should be committed!

damien78 commented:
@FestusHagen : no pb, just committed the style changes, and fixed the copyright dates. Fortunately the LGPL header was correct :)

scotchi commented on taglib/dsdiff/dsdifffile.cpp:
```diff
@@ -0,0 +1,761 @@
+    if(childChunksToUpdate.size() > 0) {
+      childChunksToUpdate[0].offset = d->chunks[d->childChunkIndex[DIINChunk]].offset + 12;
+      for(unsigned int i = 1; i < childChunksToUpdate.size(); i++)
+        childChunksToUpdate[i].offset = childChunksToUpdate[i - 1].offset + 12 + childChunksToUpdate[i - 1].size + childChunksToUpdate[i - 1].padding;
```
Lines should be kept at approximately 80 characters (with some flexibility up to 95-ish)

ufleisch commented:
I stumbled upon a DSF file which could not be read by the current code because the file has a few more bytes than the file size of the DSF header, therefore the code at line 193

```
  // File is malformed or corrupted
  if(d->fileSize != length()) {
    debug("DSF::File::read() -- File is corrupted wrong length");
    setValid(false);
    return;
  }
```

will make it impossible to save any tags or fix the file. I would suggest to change the code to

```
@@ -188,13 +188,13 @@ void DSF::File::read(bool readProperties, Properties::ReadStyle propertiesStyle)
     return;
   }
 
   d->fileSize = readBlock(8).toLongLong(false);
 
   // File is malformed or corrupted
-  if(d->fileSize != length()) {
+  if(d->fileSize > length()) {
     debug("DSF::File::read() -- File is corrupted wrong length");
     setValid(false);
     return;
   }
 
   d->metadataOffset = readBlock(8).toLongLong(false);
```

This should not cause problems when looking at how `d->fileSize` is used in the code. Removing the tag will make the tag "sane" again (`d->fileSize == length()`), changing the tag will preserve the "garbage" at the end of the file.

sbooth commented:
A file where `d->fileSize` isn't equal to `length()` violates the DSF file format specification.  I'd be hesitant to do anything to a non-compliant file because I'd worry that the audio could become corrupted.  Do you know which tool creates files like this?

ufleisch commented:
A Kid3 user came with the problem file, he seems to use MusicBrainz Picard with a DSD tagging Plugin and Kid3, see the whole story at https://sourceforge.net/p/kid3/bugs/166/#8d6b.

sbooth commented:
This PR is continued (superseded?) by #878
